---
title: "ALJEbinf: bioinformatics tools for the Engelstaedter/Letten lab"
author: "Jan Engelstaedter"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ALJEbinf}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(ALJEbinf)
```

## Introduction

This package is meant as a toolbox for common bioinformatics workflows carried out within the Engelstaedter/Letten groups at The University of Queensland. This includes in particular working with DNA sequence data derived from antibiotic resistant bacteria as well as experimental evolution studies. Currently, three kinds of tasks are supported that will be discussed below: 1) obtaining and working with "coordinates", i.e. corresponding nucleotide or aminoacid positions within homologous sequences, 2) working with tables of observed mutations, and 3) calling variants in Sanger sequencing data.

## Coordinates

Suppose we have DNA sequences of homologous genes from different species of bacteria. We may then be interested in which nucleotide or amino acid position in one species corresponds to which position in another species. As an example, let us consider the gene *rpoB* from three species: *Escherichia coli*, *Acinetobacter baylyi*, and *Mycobacterium tuberculosis*. We can first load the required packages and then import the three sequences from a fasta file:

```{r, eval=FALSE}
library(Biostrings)
library(dplyr)
library(ALJEbinf)
seqs <- readDNAStringSet("./data/rpoB_references.fasta")
seqs
```

```{r, echo=FALSE, message=FALSE}
library(Biostrings)
library(dplyr)
library(ALJEbinf)
seqs <- readDNAStringSet("../inst/extdata/rpoB_references.fasta")
seqs
```

We can now calculate the "coordinates" between two of the species, let's say *E. coli* and *M. tuberculosis*: 

```{r}
coords <- getCoordinates(seqs[["rpoB_Mycobacterium_tuberculosis_H37Rv"]],
                         seqs[["rpoB_Escherichia_coli_MG1655"]])
```

These coordinates take the form of a data frame with two columns: `posFocal`, the position of the first, focal sequence (in this case *M. tuberculosis*), and `posRef`, the corresponding position of the reference sequence (in this case *E. coli*):

```{r}
coords[1:50, ]
```

This table indicates that the first 27 nucleotides align (same position), but then there is an insertion in the *M. tuberculosis* sequence relative to the *E. coli* sequence. Note that the alignment unpinning the coordinates is based a protein alignment even though the positions are given as nucleotide positions.

Using these coordinates, we can now translate nucleotide positions from *M. tuberculosis* to *E.coli*:

```{r}
translateCoordinate(10, coords)
translateCoordinate(30, coords)
translateCoordinate(50, coords)
```

We can also translate into the opposite direction:
```{r}
translateCoordinate(10, coords, direction = "RefToFocal")
translateCoordinate(29, coords, direction = "RefToFocal")
```

Instead of nucleotide positions, we can also translate between aminoacid (or codon) positions, again in either direction:

```{r}
translateCoordinate(5, coords, AAinput = TRUE, AAoutput = TRUE)
translateCoordinate(10, coords, AAinput = TRUE, AAoutput = TRUE)
translateCoordinate(5, coords, direction = "RefToFocal", AAinput = TRUE, AAoutput = TRUE)
translateCoordinate(10, coords, direction = "RefToFocal", AAinput = TRUE, AAoutput = TRUE)
```

Finally, we can use the function `getAllCoordinates` to calculate coordinates for not just one pair of sequences but for all sequences in a list against one particular reference sequence:

```{r}
allCoords <- getAllCoordinates(seqs, "rpoB_Escherichia_coli_MG1655")
```

Here, the second argument is the name of the sequence (within the `seqs` list of sequences) to be used as the reference. The output is a list of coordinate tables that each can be used as above to translate between coordinates.

## Mutation tables

In this section we will cover the `fillMutationsTable` function, which is designed to bring data for mutations (particularly from mutant screens) into a standardised format. The focus is very much on point mutations and currently the function does not deal well with other mutations such as indels.

The starting point is a table that must have exactly the following columns:

```{r, echo = FALSE, message = FALSE}
muts <- readr::read_csv("../inst/extdata/mutations.csv",
                        col_types = readr::cols(.default = "c"))
columns <- data.frame(Column = names(muts),
                      Description = c("Name of species",
                                      "Strain name (if any)",
                                      "Genotype name (if any)",
                                      "What gene is the mutation in?",
                                      "Nucleotide position within this gene",
                                      "Nucleotide position in E. coli coordinates",
                                      "Original nucleotide in the wildtype",
                                      "Nucleotide in mutant",
                                      "Mutation at nucleotide level, e.g., 'A -> C'",
                                      "Mutation name in standard format, e.g., 'rpoB_A1576C'",
                                      "Mutation name in the same format but with E. coli coordinates",
                                      "Codon in wildtype",
                                      "Mutated codon",
                                      "Mutation at codon level, e.g., 'AGG -> CGG'",
                                      "Amino acid position",
                                      "Amino acid position in E. coli coordinates",
                                      "Original amino acid in wildtype",
                                      "Amino acid in mutant",
                                      "Mutation at amino acid level, e.g., 'S -> F'",
                                      "Mutation name in standard format, e.g. 'rpoB_S566F'",
                                      "Mutation name in same format but with E. coli coordinates"))
```
```{r}
knitr::kable(columns)
```

(The table could contain additional columns as well though.)

Suppose we have such a table of mutations in the rpoB gene from the same three species as above:

```{r}
muts
```

Since this table has too many columns to display at once, we will break it down into groups:

```{r}
muts |> select(Species:Nt_mutation)
muts |> select(Nt_OtoM:Codon_OtoM)
muts |> select(AA_pos:AA_mut_name_Ecoli)
```

We can see that this table is quite sparse (many of the entries are missing), and that it is not very standardised at this point (different mutations are specified in different ways). This is where the `fillMutationsTable` function can help. To run it, we need to supply it with 1) the mutations table, 2) a table specifying reference sequences, 3) the reference sequences themselves, and 4) a list of coordinates. We already have 1) as well as 3) and 4) from the previous section.

```{r, echo = FALSE, message = FALSE}
refs <- readr::read_csv("../inst/extdata/reference_sequences.csv")
```

The table specifying the reference sequences needs to take the following shape:

```{r}
refs
```

This table is important to link the information about species, strains and genes in the mutations table to the appropriate reference sequences. The third column, "RefStrain" indicates whether or not the sequence can be used as a general reference for all strains of that species - if the value is FALSE then the reference will only be used if the strain name matches the one in the mutations table.

We can now call the `fillMutationsTable` function:

```{r}
filledMuts <- fillMutationsTable(muts, refs, seqs, allCoords)
```

As the result shows, this table is now completed as much as possible:

```{r}
filledMuts |> select(Species:Nt_mutation)
filledMuts |> select(Nt_OtoM:Codon_OtoM)
filledMuts |> select(AA_pos:AA_mut_name_Ecoli)
```

(The mutation in the 4th row could not be completed because no nucleotide information was available.)

We can also use the `mutationsTableSummary` function to gain some overall information about the created mutations table:

```{r}
mutationsTableSummary(filledMuts)
```

The difference between the last two statistics is that the number of unique orthologous AA mutations only counts mutations as separate that are not orthologous to existing mutations in other species (achieved using the E. coli position information).

## Calling variants in Sanger sequencing data

In this workflow, we would like to use Sanger sequencing to determine which variants are present in a set of bacterial samples, which typically could be mutants obtained in a mutant screen. We need the following data: 1) fasta files containing called sequences for all samples, 2) a sample key linking fasta files to sample information, and 3) one or more reference sequences for comparison with wildtype bacteria.

The sample key is a data frame that needs to have (at least) the following columns:

```{r, echo = FALSE, message = FALSE}
sampleKey <- readr::read_csv("../inst/extdata/sample_key.csv",
                             col_types = readr::cols(.default = "c"))
columns <- data.frame(Column = names(sampleKey),
                      Description = c("A unique identifier for each species",
                                      "Name of species",
                                      "Strain name (if any)",
                                      "What gene has been amplified?",
                                      "Name of the primer used in PCR reaction and sequencing",
                                      "Name of fasta file (without extension)",
                                      "Whether or not the sequencing was successful; TRUE or FALSE",
                                      "ID of the reference sequence"))
```

```{r}
knitr::kable(columns)
```

Suppose we have our fasta files within a folder "./data/fasta", a fasta file of reference sequences (same rpoB sequences as in the previous section), and a file "./data/sample_key.csv" containing the following following sample key:

```{r}
sampleKey
```

We can now use the function `callSangerVariants_fasta` to determine which variants there are in our Sanger data:

```{r, eval=FALSE}
variants <- callSangerVariants_fasta(sampleKey_file = "./data/sample_key.csv",
                                     fasta_path = "./data/fasta/",
                                     reference_files = "./data/rpoB_references.fasta")
```

```{r, echo=FALSE}
variants <- callSangerVariants_fasta(sampleKey_file = "../inst/extdata/sample_key.csv",
                                     fasta_path = "../inst/extdata/",
                                     reference_files = "../inst/extdata/rpoB_references.fasta")
```

The resulting object `variants` is a list containing two data frames. The first one provides a summary of how many variants of different types were detected for each sample. The first columns of this table reflect the sample key:

```{r}
variants$summary |> select(SampleID:Reference)
```

The remaining columns show the results of the variant calling:

```{r}
variants$summary |> select(ContigLength:nDeletions)
```

The second data frame contains more detailed information about each variant:

```{r}
variants$details
```
The naming of the columns here is consistant with that used by the `fillMutationsTable` function (see previous section).
