---
title: "ALJEbinf: bioinformatics tools for the Engelstaedter/Letten lab"
author: "Jan Engelstaedter"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ALJEbinf}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(ALJEbinf)
```

## Introduction

This package is meant as a toolbox for common bioinformatics workflows carried out within the Engelstaedter/Letten groups at The University of Queensland. This includes in particular working with DNA sequence data derived from antibiotic resistant bacteria as well as experimental evolution studies. Currently, three kinds of tasks are supported that will be discussed below: 1) obtaining and working with "coordinates", i.e. corresponding nucleotide or aminoacid positions within homologous sequences, 2) working with tables of observed mutations, and 3) calling variants in Sanger sequencing data.

## Coordinates

Suppose we have DNA sequences of homologous genes from different species of bacteria. We may then be interested in which nucleotide or amino acid position in one species corresponds to which position in another species. As an example, let us consider the gene *rpoB* from three species: *Escherichia coli*, *Acinetobacter baylyi*, and *Mycobacterium tuberculosis*. We can first load the required packages and then import the three sequences from a fasta file:

```{r, eval=FALSE}
library(Biostrings)
library(ALJEbinf)
seqs <- readDNAStringSet("./data/rpoB_references.fasta")
seqs
```

```{r, echo=FALSE, message=FALSE}
library(Biostrings)
library(ALJEbinf)
seqs <- readDNAStringSet("../inst/extdata/rpoB_references.fasta")
seqs
```

We can now calculate the "coordinates" between two of the species, let's say *E. coli* and *M. tuberculosis*: 

```{r}
coords <- getCoordinates(seqs[["rpoB_Mycobacterium_tuberculosis_H37Rv"]],
                         seqs[["rpoB_Escherichia_coli_MG1655"]])
```

These coordinates take the form of a data frame with two columns: `posFocal`, the position of the first, focal sequence (in this case *M. tuberculosis*), and `posRef`, the corresponding position of the reference sequence (in this case *E. coli*):

```{r}
coords[1:30, ]
```

## Mutation tables

## Calling variants in Sanger sequencing data

