# Functions to generate "mutations tables", i.e. tables listing mutations
# in one or several sequences. Each row in such a table represents a particular
# mutation, and the fillMutationsTable function tries to make sure that the table
# is comprehensively filled.


#' Checks if a row in a mutations table indicates a point mutation
#'
#' @param muts A mutations table, as generated by `fillMutationsTable`.
#' @param k Row within that table.
#'
#' @return TRUE or FALSE
#'
#' @keywords internal
#'
isPointMutation <- function(muts, k) {
  pointMut <- TRUE

  # check nucleotides:
  if (!is.na(muts$Nt_OtoM[k])) {
    if (!stringr::str_detect(muts$Nt_OtoM[k], " -> "))
      pointMut <- FALSE
    nts <- stringr::str_split(muts$Nt_OtoM[k], pattern = stringr::fixed(" -> "))[[1]]
    if ((length(nts) != 2L) || (nchar(nts[1]) != 1L) || (nchar(nts[2]) != 1L))
      pointMut <- FALSE
  }

  # check codons:
  if (!is.na(muts$Codon_OtoM[k])) {
    if (!stringr::str_detect(muts$Codon_OtoM[k], " -> "))
      pointMut <- FALSE
    codons <- stringr::str_split(muts$Codon_OtoM[k], pattern = stringr::fixed(" -> "))[[1]]
    if ((length(codons) != 2L) || (nchar(codons[1]) != 3L) || (nchar(codons[2]) != 3L))
      pointMut <- FALSE
  }
  if (!is.na(muts$Codon_original[k])) {
    if (nchar(muts$Codon_original[k]) != 3L)
      pointMut <- FALSE
  }
  if (!is.na(muts$Codon_mutation[k])) {
    if (nchar(muts$Codon_mutation[k]) != 3L)
      pointMut <- FALSE
  }

  # check AAs:
  if (!is.na(muts$AA_OtoM[k])) {
    if (!stringr::str_detect(muts$AA_OtoM[k], " -> "))
      pointMut <- FALSE
    aas <- stringr::str_split(muts$AA_OtoM[k], pattern = stringr::fixed(" -> "))[[1]]
    if ((length(aas) != 2L) || (nchar(aas[1]) != 1L) || (nchar(aas[2]) != 1L))
      pointMut <- FALSE
  }
  return(pointMut)
}

# internal function to fill the DNA columns in a mutations table row:
fillMutationsTableRow_DNA <- function(muts, k, refSeq_ID, seqs, coordinates) {
  for(i in 1:2) { # repeat filling to make sure all columns are captured
    # fill Nt_OtoM from Nt_original+Nt_mutation:
    if (!is.na(muts$Nt_original[k]) && !is.na(muts$Nt_mutation[k])) {
      Nt_OtoM <- paste0(muts$Nt_original[k], " -> ", muts$Nt_mutation[k])
      if (is.na(muts$Nt_OtoM[k])) {
        muts$Nt_OtoM[k] <- Nt_OtoM
      } else {
        if (muts$Nt_OtoM[k] != Nt_OtoM)
          warning(paste0("Nt_OtoM column not consistent with Nt_original and Nt_mutation columns in row ", k, "."))
      }
    }

    # fill Nt_original and Nt_mutation from Nt_OtoM:
    if (!is.na(muts$Nt_OtoM[k])) {
      Nt_original <- stringr::word(muts$Nt_OtoM[k], 1)
      Nt_mutation <- stringr::word(muts$Nt_OtoM[k], 2, sep = stringr::fixed(" -> "))
      if (is.na(muts$Nt_original[k])) {
        muts$Nt_original[k] <- Nt_original
      } else {
        if (muts$Nt_original[k] != Nt_original)
          warning(paste0("Nt_original column not consistent with Nt_OtoM column in row ", k, "."))
      }
      if (is.na(muts$Nt_mutation[k])) {
        muts$Nt_mutation[k] <- Nt_mutation
      } else {
        if (muts$Nt_mutation[k] != Nt_mutation)
          warning(paste0("Nt_mutation column not consistent with Nt_OtoM column in row ", k, "."))
      }
    }

    # fill Nt_pos from Nt_pos_Ecoli:
    if (!is.na(muts$Nt_pos[k])) {
      Nt_pos_Ecoli <- translateCoordinate(muts$Nt_pos[k],
                                          coordinates[[refSeq_ID]],
                                          direction = "FocalToRef")
      if (is.na(Nt_pos_Ecoli)) {
        warning(paste0("Nt_pos (", muts$Nt_pos[k], ", in row ", k, ") predicted to not exist in E. coli."))
      }
      else if (is.na(muts$Nt_pos_Ecoli[k])) {
        muts$Nt_pos_Ecoli[k] <- Nt_pos_Ecoli
      } else {
        if (muts$Nt_pos_Ecoli[k] != Nt_pos_Ecoli)
          warning(paste0("Nt_pos_Ecoli column not consistent with Nt_pos column in row ", k, "."))
      }
    }

    # fill Nt_pos_Ecoli from Nt_pos:
    if (!is.na(muts$Nt_pos_Ecoli[k])) {
      Nt_pos <- translateCoordinate(muts$Nt_pos_Ecoli[k],
                                    coordinates[[refSeq_ID]],
                                    direction = "RefToFocal")
      if (is.na(Nt_pos)) {
        warning(paste0("Nt_pos_Ecoli (", muts$Nt_pos_Ecoli[k], ", in row ", k, ") predicted to not exist in this species."))
      }
      else if (is.na(muts$Nt_pos[k])) {
        muts$Nt_pos[k] <- Nt_pos
      } else {
        if (muts$Nt_pos[k] != Nt_pos)
          warning(paste0("Nt_pos column not consistent with Nt_pos_Ecoli column in row ", k, "."))
      }
    }

    # filling in Gene, Nt_original, Nt_pos and Nt_mutation from mut_name:
    if (!is.na(muts$Nt_mut_name[k])) {
      gene <- stringr::str_split(muts$Nt_mut_name[k], "_")[[1]][1]
      mutName <- stringr::str_split(muts$Nt_mut_name[k], "_")[[1]][2]
      Nt_pos <- as.integer(stringr::str_extract(mutName, "\\d+"))
      Nt_original <- stringr::str_extract_all(mutName, "[A-Z]+")[[1]][1]
      Nt_mutation <- stringr::str_extract_all(mutName, "[A-Z]+")[[1]][2]
      if (is.na(muts$Gene[k])) {
        muts$Gene[k] <- gene
      } else {
        if(muts$Gene[k] != gene)
          warning(paste0("Gene column not consistent with Nt_mut_name column in row ", k, ","))
      }
      if (is.na(muts$Nt_pos[k])) {
        muts$Nt_pos[k] <- Nt_pos
      } else {
        if(muts$Nt_pos[k] != Nt_pos)
          warning(paste0("Nt_pos column not consistent with Nt_mut_name column in row ", k, ","))
      }
      if (is.na(muts$Nt_original[k])) {
        muts$Nt_original[k] <- Nt_original
      } else {
        if(muts$Nt_original[k] != Nt_original)
          warning(paste0("Nt_original column not consistent with Nt_mut_name column in row ", k, ","))
      }
      if (is.na(muts$Nt_mutation[k])) {
        muts$Nt_mutation[k] <- Nt_mutation
      } else {
        if(muts$Nt_mutation[k] != Nt_mutation)
          warning(paste0("Nt_mutation column not consistent with Nt_mut_name column in row ", k, ","))
      }
    }

    # filling Nt_mut_name from Nt_pos, Nt_original & Nt_mutation:
    if (!is.na(muts$Nt_pos[k]) && !is.na(muts$Nt_original[k]) && !is.na(muts$Nt_mutation[k])) {
      mutName <- paste0(muts$Gene[k], "_",
                        muts$Nt_original[k],
                        muts$Nt_pos[k],
                        muts$Nt_mutation[k])
      if (is.na(muts$Nt_mut_name[k])) {
        muts$Nt_mut_name[k] <- mutName
      } else {
        if (muts$Nt_mut_name[k] != mutName) {
          warning(paste0("Nt_mut_name column not consistent with Nt_original, Nt_pos and Nt_mutation columns in row ", k, ","))
        }
      }
    }

    # filling in Nt_original, Nt_pos_Ecoli and Nt_mutation from mut_name_Ecoli:
    if (!is.na(muts$Nt_mut_name_Ecoli[k])) {
      gene <- stringr::str_split(muts$Nt_mut_name_Ecoli[k], "_")[[1]][1]
      mutName_Ecoli <- stringr::str_split(muts$Nt_mut_name_Ecoli[k], "_")[[1]][2]
      Nt_pos_Ecoli <- as.integer(stringr::str_extract(mutName_Ecoli, "\\d+"))
      Nt_original <- stringr::str_extract_all(mutName_Ecoli, "[A-Z]+")[[1]][1]
      Nt_mutation <- stringr::str_extract_all(mutName_Ecoli, "[A-Z]+")[[1]][2]
      if (is.na(muts$Nt_pos_Ecoli[k])) {
        muts$Nt_pos_Ecoli[k] <- Nt_pos_Ecoli
      } else {
        if(muts$Nt_pos_Ecoli[k] != Nt_pos_Ecoli)
          warning(paste0("Nt_pos_Ecoli column not consistent with Nt_mut_name_Ecoli column in row ", k, ","))
      }
      if (is.na(muts$Nt_original[k])) {
        muts$Nt_original[k] <- Nt_original
      } else {
        if(muts$Nt_original[k] != Nt_original)
          warning(paste0("Nt_original column not consistent with Nt_mut_name_Ecoli column in row ", k, ","))
      }
      if (is.na(muts$Nt_mutation[k])) {
        muts$Nt_mutation[k] <- Nt_mutation
      } else {
        if(muts$Nt_mutation[k] != Nt_mutation)
          warning(paste0("Nt_mutation column not consistent with Nt_mut_name_Ecoli column in row ", k, ","))
      }
    }

    # filling Nt_mut_name_Ecoli from Nt_pos, Nt_original & Nt_mutation:
    if (!is.na(muts$Nt_pos_Ecoli[k]) && !is.na(muts$Nt_original[k]) && !is.na(muts$Nt_mutation[k])) {
      mutName_Ecoli <- paste0(muts$Gene[k], "_",
                              muts$Nt_original[k],
                              muts$Nt_pos_Ecoli[k],
                              muts$Nt_mutation[k])
      if (is.na(muts$Nt_mut_name_Ecoli[k])) {
        muts$Nt_mut_name_Ecoli[k] <- mutName_Ecoli
      } else {
        if (muts$Nt_mut_name_Ecoli[k] != mutName_Ecoli) {
          warning(paste0("Nt_mut_name_Ecoli column not consistent with Nt_original, Nt_pos_Ecoli and Nt_mutation columns in row ", k, ","))
        }
      }
    }

    # filling in Codon_original and Codon_mutation from Nt_pos, Nt_original and Nt_mutation:
    if (!is.na(muts$Nt_pos[k]) && !is.na(muts$Nt_original[k]) && !is.na(muts$Nt_mutation[k])) {
      posWithinCodon <- ((muts$Nt_pos[k] - 1) %% 3) + 1
      codon_original <- stringr::str_sub(as.character(seqs[[refSeq_ID]]),
                                         muts$Nt_pos[k] - posWithinCodon + 1,
                                         muts$Nt_pos[k] - posWithinCodon + 3)
      codon_mutation <- codon_original
      stringr::str_sub(codon_mutation, posWithinCodon, posWithinCodon) <- muts$Nt_mutation[k]
      if (is.na(muts$Codon_original[k])) {
        muts$Codon_original[k] <- codon_original
      } else {
        if (muts$Codon_original[k] != codon_original)
          warning(paste0("Codon_original column not consistent with Nt_pos columns in row ", k, ","))
      }
      if (is.na(muts$Codon_mutation[k])) {
        muts$Codon_mutation[k] <- codon_mutation
      } else {
        if (muts$Codon_mutation[k] != codon_mutation)
          warning(paste0("Codon_mutation column not consistent with Nt_pos and Nt_mutation columns in row ", k, ","))
      }
    }

    # filling in Codon_OtoM from Codon_original and Codon_mutation:
    if (!is.na(muts$Codon_original[k]) && !is.na(muts$Codon_mutation[k])) {
      codon_OtoM <- paste(muts$Codon_original[k], "->", muts$Codon_mutation[k])
      if (is.na(muts$Codon_OtoM[k])) {
        muts$Codon_OtoM[k] <- codon_OtoM
      } else {
        if (muts$Codon_OtoM[k] != codon_OtoM)
          warning(paste0("Codon_OtoM column not consistent with Codon_Original and Codon_mutation columns in row ", k, ","))
      }
    }

    # filling in Codon_original and Codon_mutation from Codon_OtoM:
    if (!is.na(muts$Codon_OtoM[k])) {
      Codon_original <- stringr::word(muts$Codon_OtoM[k], 1)
      Codon_mutation <- stringr::word(muts$Codon_OtoM[k], 2, sep = stringr::fixed(" -> "))
      if (is.na(muts$Codon_original[k])) {
        muts$Codon_original[k] <- Codon_original
      } else {
        if (muts$Codon_original[k] != Codon_original)
          warning(paste0("Codon_original column not consistent with Codon_OtoM column in row ", k, "."))
      }
      if (is.na(muts$Codon_mutation[k])) {
        muts$Codon_mutation[k] <- Codon_mutation
      } else {
        if (muts$Codon_mutation[k] != Codon_mutation)
          warning(paste0("Codon_mutation column not consistent with Codon_OtoM column in row ", k, "."))
      }
    }

    # filling in Nt_original, Nt_mutation and Nt_pos from Codon and AA_pos:
    if (!is.na(muts$Codon_original[k]) &&
        !is.na(muts$Codon_mutation[k]) &&
        !is.na(muts$AA_pos[k])) {
      nts_original <- stringr::str_split(muts$Codon_original[k], "")[[1]]
      nts_mutation <- stringr::str_split(muts$Codon_mutation[k], "")[[1]]
      if (sum(nts_original != nts_mutation) == 1) { # one nucleotide difference?
        Nt_pos <- (3L * muts$AA_pos[k]) - 3L + which(nts_original != nts_mutation)
        Nt_original <- nts_original[nts_original != nts_mutation]
        Nt_mutation <- nts_mutation[nts_original != nts_mutation]

        if (is.na(muts$Nt_pos[k])) {
          muts$Nt_pos[k] <- Nt_pos
        } else {
          if (muts$Nt_pos[k] != Nt_pos)
            warning(paste0("Nt_pos column not consistent with codons and AA_pos position in row ", k, "."))
        }
        if (is.na(muts$Nt_original[k])) {
          muts$Nt_original[k] <- Nt_original
        } else {
          if (muts$Nt_original[k] != Nt_original)
            warning(paste0("Nt_original column not consistent with codons and AA_pos column in row ", k, "."))
        }
        if (is.na(muts$Nt_mutation[k])) {
          muts$Nt_mutation[k] <- Nt_mutation
        } else {
          if (muts$Nt_mutation[k] != Nt_mutation)
            warning(paste0("Nt_mutation column not consistent with codons and AA_pos column in row ", k, "."))
        }
      }
      else if ((sum(nts_original != nts_mutation) == 0)) {
        warning(paste0("No nucleotide differences between Codon_original and Codon_mutation in row ", k, "."))
      } else {
        warning(paste0("More than one nucleotide differences between Codon_original and Codon_mutation in row ", k, "."))
      }
    }

    # filling in Nt_original, Nt_mutation and Nt_pos from Codon and AA_pos:
    if (!is.na(muts$Codon_original[k]) &&
        !is.na(muts$Codon_mutation[k]) &&
        !is.na(muts$AA_pos_Ecoli[k])) {
      nts_original <- stringr::str_split(muts$Codon_original[k], "")[[1]]
      nts_mutation <- stringr::str_split(muts$Codon_mutation[k], "")[[1]]
      if (sum(nts_original != nts_mutation) == 1) { # one nucleotide difference?
        Nt_pos_Ecoli <- (3L * muts$AA_pos_Ecoli[k]) - 3L + which(nts_original != nts_mutation)
        Nt_original <- nts_original[nts_original != nts_mutation]
        Nt_mutation <- nts_mutation[nts_original != nts_mutation]

        if (is.na(muts$Nt_pos_Ecoli[k])) {
          muts$Nt_pos_Ecoli[k] <- Nt_pos_Ecoli
        } else {
          if (muts$Nt_pos_Ecoli[k] != Nt_pos_Ecoli)
            warning(paste0("Nt_pos_Ecoli column not consistent with codons and AA_pos position in row ", k, "."))
        }
        if (is.na(muts$Nt_original[k])) {
          muts$Nt_original[k] <- Nt_original
        } else {
          if (muts$Nt_original[k] != Nt_original)
            warning(paste0("Nt_original column not consistent with codons and AA_pos column in row ", k, "."))
        }
        if (is.na(muts$Nt_mutation[k])) {
          muts$Nt_mutation[k] <- Nt_mutation
        } else {
          if (muts$Nt_mutation[k] != Nt_mutation)
            warning(paste0("Nt_mutation column not consistent with codons and AA_pos column in row ", k, "."))
        }
      }
      else if ((sum(nts_original != nts_mutation) == 0)) {
        warning(paste0("No nucleotide differences between Codon_original and Codon_mutation in row ", k, "."))
      } else {
        warning(paste0("More than one nucleotide differences between Codon_original and Codon_mutation in row ", k, "."))
      }
    }
  }
  return(muts)
}


# internal function to fill the amino acid columns in a mutations table row:
fillMutationsTableRow_Protein <- function(muts, k, refSeq_ID, seqs, coordinates) {

  # translating from nucleotides to amino acids:
  if (!is.na(muts$Nt_pos[k]) &&
      !is.na(muts$Codon_original[k]) &&
      !is.na(muts$Codon_mutation[k])) {
    aa_pos <- (muts$Nt_pos[k] - 1L) %/% 3L + 1L
    aa_original <- as.character(Biostrings::translate(Biostrings::DNAString(muts$Codon_original[k]), no.init.codon = TRUE))
    aa_mutation <- as.character(Biostrings::translate(Biostrings::DNAString(muts$Codon_mutation[k]), no.init.codon = TRUE))

    if (is.na(muts$AA_pos[k])) {
      muts$AA_pos[k] <- aa_pos
    } else {
      if (muts$AA_pos[k] != aa_pos)
        warning(paste0("AA_pos column not consistent with Nt_pos column in row ", k, "."))
    }
    if (is.na(muts$AA_original[k])) {
      muts$AA_original[k] <- aa_original
    } else {
      if (muts$AA_original[k] != aa_original)
        warning(paste0("AA_original column not consistent with Codon_original column in row ", k, "."))
    }
    if (is.na(muts$AA_mutation[k])) {
      muts$AA_mutation[k] <- aa_mutation
    } else {
      if (muts$AA_mutation[k] != aa_mutation)
        warning(paste0("AA_mutation column not consistent with Codon_mutation column in row ", k, "."))
    }
  }

  # filling in AA_pos from AA_pos_Ecoli:
  if (!is.na(muts$AA_pos_Ecoli[k])) {
    aa_pos <- translateCoordinate(muts$AA_pos_Ecoli[k],
                                  coordinates[[refSeq_ID]],
                                  direction = "RefToFocal",
                                  AAinput = TRUE,
                                  AAoutput = TRUE)
    if (is.na(aa_pos)) {
      warning(paste0("AA_pos_Ecoli (", muts$AA_pos_Ecoli[k], ", in row ", k, ") predicted to not exist in this species."))
    }
    else if (is.na(muts$AA_pos[k])) {
      muts$AA_pos[k] <- aa_pos
    } else {
      if (muts$AA_pos[k] != aa_pos)
        warning(paste0("AA_pos column not consistent with AA_pos_Ecoli column in row ", k, "."))
    }
  }

  # filling in AA_pos_Ecoli from AA_pos:
  if (!is.na(muts$AA_pos[k])) {
    aa_pos_Ecoli <- translateCoordinate(muts$AA_pos[k],
                                        coordinates[[refSeq_ID]],
                                        direction = "FocalToRef",
                                        AAinput = TRUE,
                                        AAoutput = TRUE)
    if (is.na(aa_pos_Ecoli)) {
      warning(paste0("AA_pos (", muts$AA_pos[k], ", in row ", k, ") predicted to not exist in E. coli."))
    }
    else if (is.na(muts$AA_pos_Ecoli[k])) {
      muts$AA_pos_Ecoli[k] <- aa_pos_Ecoli
    } else {
      if (muts$AA_pos_Ecoli[k] != aa_pos_Ecoli)
        warning(paste0("AA_pos_Ecoli column not consistent with AA_pos column in row ", k, "."))
    }
  }

  # filling in AA_orignal and AA_mutation from AA_OtoM:
  if (!is.na(muts$AA_OtoM[k])) {
    aa_original <- stringr::str_extract(muts$AA_OtoM[k], ".+?(?= ->)")
    aa_mutation <- stringr::str_extract(muts$AA_OtoM[k], "(?<=-> ).*$")
    if (is.na(muts$AA_original[k])) {
      muts$AA_original[k] <- aa_original
    } else {
      if (muts$AA_original[k] != aa_original)
        warning(paste0("AA_original column not consistent with AA_OtoM column in row ", k, "."))
    }
    if (is.na(muts$AA_mutation[k])) {
      muts$AA_mutation[k] <- aa_mutation
    } else {
      if (muts$AA_mutation[k] != aa_mutation)
        warning(paste0("AA_mutation column not consistent with AA_OtoM column in row ", k, "."))
    }
  }

  # filling in AA_OtoM from AA_orignal and AA_mutation:
  if (!is.na(muts$AA_original[k]) &&
      !is.na(muts$AA_mutation[k])) {
    aa_OtoM <- paste(muts$AA_original[k], "->", muts$AA_mutation[k])
    if (is.na(muts$AA_OtoM[k])) {
      muts$AA_OtoM[k] <- aa_OtoM
    } else {
      if (muts$AA_OtoM[k] != aa_OtoM)
        warning(paste0("AA_OtoM column not consistent with AA_original and AA_mutation columns in row ", k, "."))
    }
  }

  # filling in AA_mut_name from AA_pos, AA_original and AA_mutation:
  if (!is.na(muts$AA_original[k]) &&
      !is.na(muts$AA_mutation[k]) &&
      !is.na(muts$AA_pos[k])) {
    aa_mut_name <- paste0(muts$Gene[k], "_",
                          muts$AA_original[k],
                          muts$AA_pos[k],
                          muts$AA_mutation[k])
    if (is.na(muts$AA_mut_name[k] )) {
      muts$AA_mut_name[k] <- aa_mut_name
    } else {
      if (muts$AA_mut_name[k] != aa_mut_name)
        warning(paste0("AA_mut_name column not consistent with AA_pos, AA_original, AA_mutation columns in row ", k, "."))
    }
  }

  # filling in Gene, AA_original, AA_pos and AA_mutation from AA_mut_name:
  # To be completed!

  # filling in AA_mut_name_Ecoli from AA_pos_Ecoli, AA_original and AA_mutation:
  if (!is.na(muts$AA_original[k]) &&
      !is.na(muts$AA_mutation[k]) &&
      !is.na(muts$AA_pos_Ecoli[k])) {
    aa_mut_name_Ecoli <- paste0(muts$Gene[k], "_",
                                muts$AA_original[k],
                                muts$AA_pos_Ecoli[k],
                                muts$AA_mutation[k])
    if (is.na(muts$AA_mut_name_Ecoli[k] )) {
      muts$AA_mut_name_Ecoli[k] <- aa_mut_name_Ecoli
    } else {
      if (muts$AA_mut_name_Ecoli[k] != aa_mut_name_Ecoli)
        warning(paste0("AA_mut_name_Ecoli column not consistent with AA_pos_Ecoli, AA_original, AA_mutation columns in row ", k, "."))
    }
  }
  return(muts)
}


#' Fill in missing data in a mutations table.
#'
#' This function tries to infer missing data in a mutations table,
#' including mutations at the amino acid level when only the nucleotide
#' level information is present, explicit positions of mutations where
#' this data is only implicitly available, and positions in reference (E.coli)
#' coordinates.
#' The result is a standardised table that can then more readily be analysed.
#'
#' @param muts An (incomplete) mutations table.
#' @param refs A table specifying the species and names for all reference sequences.
#' @param seqs Reference sequences as a DNAStringSet object, with names as specified in `refs`.
#' @param coordinates A list of coordinates, with names corresponding to seqs.
#'
#' @return A muts table that is (hopefully) more complete than the input table.
#' @export
#'
fillMutationsTable <- function(muts, refs, seqs, coordinates) {
  # convert positions into numbers:
  # (needs to be improved if mutations other than point mutations are to be included)
  muts$Nt_pos <- as.integer(muts$Nt_pos)
  muts$Nt_pos_Ecoli <- as.integer(muts$Nt_pos_Ecoli)
  muts$AA_pos <- as.integer(muts$AA_pos)
  muts$AA_pos_Ecoli <- as.integer(muts$AA_pos_Ecoli)
  #muts$MIC_mgPerL <- as.double(muts$MIC_mgPerL)

  # add strainID columns:
  muts <- dplyr::mutate(muts, Strain_ID = gsub(" ", "_", paste(Species, Strain)))
  muts <- dplyr::mutate(muts, RefSeq_ID = NA)
  refs <- dplyr::mutate(refs, Strain_ID = gsub(" ", "_", paste(Species, Strain)))

  genes <- unique(muts$Gene)
  strainIDs <- unique(muts$Strain_ID)
  for(i in 1:length(genes)) {
    for(j in 1:length(strainIDs)) {
      refSeq_ID <- NA
      # check if there are any mutations in that strain j and gene i:
      if ((muts |>
           dplyr::filter(Gene == genes[i], Strain_ID == strainIDs[j]) |>
           nrow()) > 0) {
        geneStrain <- paste(genes[i], strainIDs[j], sep = "_")
        cat(paste0("Working on gene_species_strain ", geneStrain, ".\n"))
        # finding the appropriate reference sequence for the gene:
        if (geneStrain %in% names(seqs)) {  # specific strain has reference
          refSeq_ID <- geneStrain
          cat("  -> Specific reference for this strain found.\n")
        } else {
          species <- paste(stringr::str_split(strainIDs[j], "_")[[1]][1:2],
                           collapse = " ")
          filteredRefs <- dplyr::filter(refs,
                                        Gene == genes[i] & Species == species & RefStrain)
          if ((nrow(filteredRefs) == 1) && (filteredRefs$FASTA_name[1] %in% names(seqs))) {
            refSeq_ID <- filteredRefs$FASTA_name[1]
            cat(paste0("  -> No specific reference found, but used ", filteredRefs$FASTA_name[1], " instead.\n"))
          } else {
            cat("  -> No reference sequence found.\n")
          }
        }
      }
      if (!is.na(refSeq_ID)) {
        for(k in 1:nrow(muts)) {
          if ((muts$Gene[k] == genes[i]) && (muts$Strain_ID[k] == strainIDs[j])) {
            muts$RefSeq_ID[k] <- refSeq_ID
            if (isPointMutation(muts, k)) {
              muts <- fillMutationsTableRow_DNA(muts, k, refSeq_ID, seqs, coordinates)
              muts <- fillMutationsTableRow_Protein(muts, k, refSeq_ID, seqs, coordinates)
            } else {
              warning(paste0("Row ", k, " in mutation table does not appear to represent a single point mutation; this column has been skipped."))
            }
          }
        }
      }
    }
  }
  cat("\n")
  mutationsTableSummary(muts)
  return(muts)
}

#' Print a summary of a mutations table
#'
#' @param muts A mutations table, as created using the fillMutationsTable function.
#'
#' @export
#'
mutationsTableSummary <- function(muts) {
  cat("Summary of mutations table:\n")
  cat("  Number of species:",
      muts |> dplyr::pull(Species) |> unique() |> length(), "\n")
  cat("  Genes with mutations:",
      paste(muts |>
              dplyr::pull(Gene) |>
              unique(),
            collapse = ", "),
      "\n")
  cat("  Total number of recorded mutations:", nrow(muts), "\n")
  cat("  Number of unique AA mutations across species:",
      muts |>
        dplyr::filter(!is.na(AA_mut_name)) |>
        dplyr::group_by(Species, AA_mut_name) |>
        dplyr::summarise(n = dplyr::n(), .groups = "drop") |>
        dplyr::pull(n) |>
        sum(),
      "\n")
  cat("  Number of unique orthologous AA mutations:",
      muts |>
        dplyr::filter(!is.na(AA_mut_name)) |>
        dplyr::pull(AA_mut_name_Ecoli) |>
        unique() |>
        length(),
      "\n")

  missingRefs <- muts |>
    dplyr::filter(is.na(RefSeq_ID)) |>
    dplyr::select(Gene, Strain_ID) |>
    dplyr::distinct(Gene, Strain_ID) |>
    tidyr::unite(Gene_Strain, Gene, Strain_ID, sep = "_") |>
    dplyr::pull(Gene_Strain)
  if (length(missingRefs) > 0) {
    cat("\n")
    cat("No reference found for the following Gene_Species_Strain combinations:\n")
    cat(paste(" ", missingRefs, sep = "", collapse = "\n"))
    cat("\n\n")
  }
}
